[Rebase & FF][FF-A] Adding FDT based MemoryAllocationLib and ServicesTableLib for Secure Partitions

kuqin12 opened the pull request:
## Description

This change adds a new library instance to support heap allocation management for secure partitions.

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [ ] Backport to release branch?

## How This Was Tested

This was tested on feature/ffa_enablement branch with QEMU SBSA platform.

## Integration Instructions

- The platform needs to link this new library as the memory allocation library for secure partition builds.
- In the secure partition DTS, adding a memory region for heap will allow the memory allocation library to allocate memory from this region.

Raymond-MS commented on ArmPkg/Library/SecurePartitionMemoryAllocationLib/Page.c:
```diff
@@ -0,0 +1,399 @@
+    }
+  }
+
+  return ~Address;
```
This seems like an odd return value, what was the decision behind returning the inverted address and not -1 like in other functions?

kuqin12 replied:
This is probably a quick change to the input value. The important characteristic is that upon such failure, the return has to be different than the input value. A constant value will not work. We can do +1, -1 or other for the same effect.

Raymond-MS commented on ArmPkg/Library/SecurePartitionMemoryAllocationLib/Page.c:
```diff
@@ -0,0 +1,399 @@
+{
+  UINTN  RequestedAddress;
+
+  if ((MemoryType != EfiRuntimeServicesCode) &&
```
This parameter is only checked here but not used anywhere else in the function, why are we passing it in?

kuqin12 replied:
This is just saying that we will only allow runtime type.

Raymond-MS commented on ArmPkg/Library/SecurePartitionMemoryAllocationLib/Pool.c:
```diff
@@ -0,0 +1,329 @@
+  FreePoolHdr->Header.Available = TRUE;
+  ASSERT (PoolIndex < MAX_POOL_INDEX);
+  InsertHeadList (&mMmPoolLists[PoolIndex], &FreePoolHdr->Link);
+  return EFI_SUCCESS;
```
Any reason for this function to return a status if it will always return SUCCESS? We should just be able to make it a void return.

kuqin12 replied:
Actually a nice call. We probably should convert these asserts into error codes instead.

kuqin12 replied:
Fixed.

Raymond-MS commented on ArmPkg/Library/SecurePartitionMemoryAllocationLib/SecurePartitionMemoryAllocationLib.c:
```diff
@@ -0,0 +1,1051 @@
+{
+  EFI_STATUS  Status;
+
+  ASSERT (Pages != 0);
```
I see we do an assert on pages to verify it is a non-zero value, why don't we also check if Buffer is non-null? That or we could at least just return immediately if so.

kuqin12 replied:
Updated.

Raymond-MS commented on ArmPkg/Library/SecurePartitionMemoryAllocationLib/SecurePartitionMemoryAllocationLib.c:
```diff
@@ -0,0 +1,1051 @@
+{
+  EFI_STATUS  Status;
+
+  ASSERT (Pages != 0);
```
Same question here about checking Buffer. :)

kuqin12 replied:
Hmm, can we really rule out the Buffer never being null? It is unorthodox. But the actual usage will depend on what regions this SP will have access to?

Raymond-MS replied:
It is entirely possible that buffer will never be NULL but it never hurts to put in safe code to make sure we aren't doing something we don't want. We check Pages so I don't see why we shouldn't also check Buffer. It should always be the case that if Pages !=0 then Buffer != NULL.

kuqin12 replied:
Updated.

Raymond-MS commented on ArmPkg/Library/SecurePartitionMemoryAllocationLib/SecurePartitionMemoryAllocationLib.c:
```diff
@@ -0,0 +1,1051 @@
+  VOID  *NewBuffer;
+
+  NewBuffer = InternalAllocateZeroPool (PoolType, NewSize);
+  if ((NewBuffer != NULL) && (OldBuffer != NULL)) {
```
This is just a question really, what happens if OldBuffer is not passed in (since it is marked OPTIONAL), does it get defaulted to NULL?

kuqin12 replied:
Oh, optional means it could be NULL when caller invokes this function. Not that caller does not need to pass in anything for this argument. If I understand you correctly?

Raymond-MS commented on ArmPkg/Library/SecurePartitionMemoryAllocationLib/SecurePartitionMemoryAllocationLib.c:
```diff
@@ -0,0 +1,1051 @@
+
+**/
+VOID *
+InternalAllocateAlignedPages (
```
Are "Internal" functions not supposed to be marked STATIC or are they not really internal? Just a question. :)

kuqin12 replied:
There is no strict definition of "Internal" here. But in this case, it means it might be used outside of this file but should stay in the module.

apop5 commented on ArmPkg/Library/SecurePartitionEntryPoint/StandaloneMmCoreEntryPoint.c:
```diff
@@ -423,12 +422,12 @@ ModuleEntryPoint (
   mUseOnlyFfaAbis = UseOnlyFfaAbis;
 
-  ProcessLibraryConstructorList (NULL, NULL);
+  ProcessLibraryConstructorList (DtbAddress, NULL);
```
I have concerns about how ImageHandle is being taken to pass DtbAddress.

Taking ImageHandle and using it in this manor seems like we are going to cause more problems down the line when we encounter a library constructor that attempts to use the passed values as they are stated in the prototypes.

kuqin12 replied:
Yeah. I was thinking about the same. The hard part is that this FDT address will be automatically populated by the loader in the input parameters at the entrypoint function. We could save it to a global variable and have others access it that way. But since I imagine this will be used in other libraries as well, using an extern will be a bad pattern. Any suggestion is welcome.

apop5 replied:
summarizing our offline conversation:

UefiBootServicesTableLib and UefiRuntimeServicesTableLib create gBS and gRT global variables, and their implementations are provided by the library instance.
Would doing something similar for DtbAddress be acceptable?
This entry point can still access it directly, but to allow library constructors to access it, they would consume the library instance directly.

apop5 replied:
Changes look good to me.

kuqin12 committed:
ArmPkg: SecurePartitionServicesTableLib: Introducing services table lib

This change adds a new library class SecurePartitionServicesTableLib for
secure partitions library instances to share common resources within the
partition environment.

The entrypoint is updated to pair with the change to initialize the
table before dispatching the driver entrypoint.

Co-authored-by: Aaron Pop <aaronpop@microsoft.com>
Signed-off-by: Kun Qin <kun.qin@microsoft.com>

apop5 commented on ArmPkg/Library/SecurePartitionServicesTableLib/SecurePartitionServicesTableLib.inf:
```diff
@@ -0,0 +1,39 @@
+## @file
+# Module entry point library for DXE core.
+#
+# Copyright (c) 2017 - 2021, Arm Ltd. All rights reserved.<BR>
```
```suggestion
# Copyright (c) Microsoft Corporation.

```

kuqin12 replied:
Thanks, updated.

apop5 commented on ArmPkg/Library/SecurePartitionServicesTableLib/SecurePartitionServicesTableLib.inf:
```diff
@@ -0,0 +1,39 @@
+## @file
+# Module entry point library for DXE core.
```
```suggestion
# Secure Partition System Table library implementation.

```

kuqin12 replied:
Updated.

apop5 commented on ArmPkg/Library/SecurePartitionEntryPoint/StandaloneMmCoreEntryPoint.c:
```diff
@@ -311,7 +302,7 @@ GetSpManifest (
   return EFI_NOT_FOUND;
 }
-
+extern volatile BOOLEAN loop;
```
left over?

kuqin12 replied:
Yep, removed.

kuqin12 committed:
ArmPkg: SecurePartitionEntryPoint: Initialize gSpst

This change adds the code to initialize gSpst in secure partition
entrypoint to be used for other library instances.

Signed-off-by: Kun Qin <kun.qin@microsoft.com>

kuqin12 committed:
ArmPkg: SecurePartitionMemoryAllocationLib: Adding a FDT based heap lib

This changes adds a MemoryAllocationLib support for secure partitions.
The change is based off the memory manager from Standalone MM core.

The secure partition needs to specify the heap base address and size in
the partition manifest to initialize the heap manager during runtime.

Signed-off-by: Kun Qin <kun.qin@microsoft.com>
